### ISSUE TITLE
ATProto: detect and handle videos that are over Bluesky's 60s limit

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#1353]
Related to #1348: [Bluesky currently only supports short videos, 60 seconds or less.](https://bsky.social/about/blog/09-11-2024-video) It'd be nice to detect videos that are too long and include a `[Video]` text suffix, like we used to do, instead of bridging them anyway and having the client show that they failed:

<img width="659" alt="image" src="https://github.com/user-attachments/assets/3121cc27-10c7-4a7b-a760-6be6a324ad73">

This is technically nontrivial though. I'd need to buffer the entire video in memory, add decoding logic for multiple different formats, who knows what else. I'm unlikely to prioritize this myself, but I'm open to merging a PR if it's reasonable!

**Comment by qazmlp** [comment]
[ffprobe](https://ffmpeg.org/ffprobe.html) would almost certainly work.
It also takes URLs as input and likely doesn't need to download the entire file.

I'm not sure how well that works with your deployment model, though.

**Comment by snarfed** [comment]
Sadly yeah, I generally need to avoid command line and filesystem-based tools.

**Comment by Daft-Freak** [comment]
This on the right track? (Yeah I know, magic 60 second number)
```diff
diff --git a/arroba/datastore_storage.py b/arroba/datastore_storage.py
index 4f1878f..5a07ec1 100644
--- a/arroba/datastore_storage.py
+++ b/arroba/datastore_storage.py
@@ -307,6 +307,10 @@ class AtpRemoteBlob(ndb.Model):
     width = ndb.IntegerProperty()
     height = ndb.IntegerProperty()
 
+    # only populated if mime_type is video/*
+    # used to enforce maximum duration
+    duration = ndb.IntegerProperty()
+
     created = ndb.DateTimeProperty(auto_now_add=True)
     updated = ndb.DateTimeProperty(auto_now=True)
 
@@ -339,10 +343,15 @@ class AtpRemoteBlob(ndb.Model):
             if max_size and size > max_size:
                 raise ValidationError(f'{url} Content-Length {size} is over {name} blob maxSize {max_size}')
 
+        def validate_duration(duration):
+            if duration and duration > 60000:
+                raise ValidationError(f'{url} duration {duration / 1000} is over 60s')
+
         assert url
         blob = cls.get_by_id(url)
         if blob:
             validate_size(blob.size)
+            validate_duration(blob.duration)
             server.validate_mime_type(blob.mime_type, accept_types, name=url)
             return blob
 
@@ -390,6 +399,7 @@ class AtpRemoteBlob(ndb.Model):
                     track = media_info.video_tracks[0]
                     blob.width = track.width
                     blob.height = track.height
+                    blob.duration = track.duration
             except (OSError, RuntimeError) as e:
                 logger.info(e)
 
@@ -399,6 +409,8 @@ class AtpRemoteBlob(ndb.Model):
         # do this after storing blob so that we don't re-download it next time.
         validate_size(len(resp.content))
 
+        validate_duration(blob.duration)
+
         return blob
 
     def as_object(self):
```

Though I suppose there's another issue here in that invalid videos just get dropped...

**Comment by snarfed** [comment]
@Daft-Freak that looks exactly right!

Minor: one other thing I tend to do when I implement external reqt's like this is to link to where they're defined or documented, in a comment. Bluesky doesn't have great docs around its limits like this that aren't in lexicons; the best I can find for this one is https://bsky.social/about/blog/09-11-2024-video 😐

**Comment by snarfed** [comment]
Another minor nit: we'd probably want to mention in the docstring's `Raises` section that it raises `ValidationError` for this even though technically it's not a lexicon schema validation error. Seems ok to me though.

**Comment by Daft-Freak** [comment]
https://github.com/snarfed/arroba/pull/49 + https://github.com/snarfed/granary/pull/889?

**Comment by snarfed** [comment]
@Daft-Freak yes! Thanks for the nudge, and thank you again for the contribution. And it's working, congrats! https://indieweb.social/@snarfed/114049210341490927 => https://bsky.app/profile/fedi.test.snarfed.org

(...despite Mastodon somehow converting or interpreting the video as a gif?! 🤷 ah well, no matter, that's not on us)

**Comment by Daft-Freak** [comment]
I think Mastodon just calls any video without audio a "GIF"... (The [original post](https://fosstodon.org/@Daft_Freak/113477367229501051) is also like that)

Honestly I'm a little surprised at how much I've got done here. Hopefully this one helps me the next time I forget about the limit 😆
